<h1 align=center><font size=5>ECON-UB. 251</font></h1>
<h1 align=center><font size=5>Econometrics I</font></h1>
<h1 align=center><font size=5>Problem Set 4</font></h1>

<h4>Instructions</h4>
<p>Problem Set 4 is due on Sunday, Dec 8, at 6pm. Please submit your Jupyter Notebook file on NYU Classes Assignment. Make sure you include your name and Net ID in the title of your notebook.</p>
<p>You are required to type your answer in the column below each of the question. Remember to execute your code (Shift+Enter). You can insert cell and use #comments if necessary. 

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1>Question 1</h1>

<font size=2>Load data via link below and store in the dataframe.</font>

In [1]:
pip install linearmodels

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
df = pd.read_csv('https://github.com/MariaMeng22/econometric-001/raw/master/voucher.csv')
df.head()

studyid  black  hispanic  female  appyear  mnce  select  choice  selectyrs  \
0       21      1         0       1       90    44       1       0          4   
1       26      1         0       0       90    46       1       1          4   
2       30      1         0       0       90    20       1       1          4   
3       31      1         0       1       90    36       1       1          4   
4       33      1         0       0       90    32       1       1          4   

   choiceyrs  mnce90  selectyrs1  selectyrs2  selectyrs3  selectyrs4  \
0          1     NaN           0           0           0           1   
1          4     NaN           0           0           0           1   
2          4     NaN           0           0           0           1   
3          4     NaN           0           0           0           1   
4          4     NaN           0           0           0           1   

   choiceyrs1  choiceyrs2  choiceyrs3  choiceyrs4  
0           1           0           0           0  
1           0           0           0           1  
2           0           0           0           1  
3           0           0           0           1  
4           0           0           0           1

Consider the following (structual model):
$$ mnce = \beta_0 + \beta_1 choiceyrs + \beta_2 black + \beta_3 hispanic + \beta_4 female + \epsilon $$
$mnce$ is the outcome of interest, the student's percentile score on a math test administered in 1994;  
$choiceyrs$ indicates the number of years from 1991 to 1994 that a student attended a choice school, ranges from 0 to 4;  
$black = 1$ if African-American;  
$hispanic = 1$ if Hispanic;  
$female = 1$ if female  

(1) Estimate the model above and interpret the coefficient on attending a choice school($choiceyrs$).


The concern is that $choiceyrs$ is not exogenous. Rouse(1998) uses $selectyrs$ as an instrumental variable for $choiceyrs$.The variable $selectyrs$ is the number of years a student who was selected for a voucher. If the student applied for the program in 1990 and received a voucher then $selectyrs = 4$; if he or she applied in 1991 and received a voucher then $selectyrs = 3$; and so on. Now we follow Rouse's method to construct the instrument.   

Use the `statsmodels` package to answer (2) - (4)  
(2) Estimate the "first stage" equation, that is, regress the endogenous variable on the explanatory variables and instrument
$$choiceyrs = \alpha_0 + \alpha_1 black + \alpha_2 hispanic + \alpha_3 female + \alpha_4selectyrs + v$$
and print out the results summary. Add a new column $choiceyrshat$ to store fitted values for $choiceyrs$ and then print out the description. Is the relationship between the endogenous and IV strong? Positive or negative?  

(3) What must we assume for $selectyrs$ to serve as valid instruments? Does your estimate from (2) support $selectyrs$?  

(4) Estimate the "second stage equation" by replacing the endogenous variable with the fitted values, $\widehat{choiceyrs}$ from first stage (which in dataset is $choiceyrshat$)
$$ mnce = \beta_0 + \beta_1 \widehat{choiceyrs} + \beta_2 black + \beta_3 hispanic + \beta_4 female + \epsilon $$
print out the results summary. Compare the coefficient values of $\widehat{choiceyrs}$ to what you got in (1) and comment.    

(5) Now we introduce the `linearmodels` package to replicate the 2SLS estimation above by using the command `IV2SLS` and using $selectyrs$ as the IV for $choiceyrs$. 


Here is a link for you to learn how to conduct IV estimation with this package. https://bashtage.github.io/linearmodels/doc/iv/examples/basic-examples.html

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 1</h3>

In [3]:
import statsmodels.formula.api as smf

In [4]:
#ols
res_ols = smf.ols('mnce ~ choiceyrs + black + hispanic + female', data=df).fit(cov_type = 'HC3')
print(res_ols.summary())

                            OLS Regression Results                            
Dep. Variable:                   mnce   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.083
Method:                 Least Squares   F-statistic:                     20.11
Date:                Sat, 07 Dec 2019   Prob (F-statistic):           6.15e-16
Time:                        12:39:15   Log-Likelihood:                -4367.3
No. Observations:                 990   AIC:                             8745.
Df Residuals:                     985   BIC:                             8769.
Df Model:                           4                                         
Covariance Type:                  HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     57.1219      1.888     30.252      0.0

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
On average an additional year in a choice school will decrease his or her math percentile point by 0.5652 point, holding others constant. 

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 2</h3>

In [5]:
res_s1 = smf.ols('choiceyrs ~ selectyrs + black + hispanic + female', data=df).fit(cov_type = 'HC3')
print(res_s1.summary())

                            OLS Regression Results                            
Dep. Variable:              choiceyrs   R-squared:                       0.790
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     486.6
Date:                Sat, 07 Dec 2019   Prob (F-statistic):          1.79e-231
Time:                        12:39:15   Log-Likelihood:                -856.97
No. Observations:                 990   AIC:                             1724.
Df Residuals:                     985   BIC:                             1748.
Df Model:                           4                                         
Covariance Type:                  HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.0191      0.028     -0.672      0.5

In [6]:
df['fitted'] = res_s1.fittedvalues
df['fitted'].describe()

count    990.000000
mean       1.022222
std        1.116176
min       -0.019054
25%        0.013304
50%        0.773808
75%        2.300672
max        3.110062
Name: fitted, dtype: float64

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
"First stage" equation: 
$choiceyrs = -0.0191 + 0.0294 black + 0.0430 hispanic + 0.0324 female + 0.7634 selectyrs + v$, p-value is 0.000, which is less than the critical p-value of 0.05 at 5% level of significance, this variable is individually significant at the 5% level of significance.  
    
The relationship between IV and endogenous variable is strong because $R^2 = 0.790$.  

The coefficient on IV is 0.7634 > 0, indicating a positive relationship between IV and the endogenous variable.

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 3</h3>
    
    
(1) z is uncorrelated with u, that is,
$$Cov(z,u) ≠ 0$$ 
(2) z is correlated with x, that is,
$$Cov(z,x) = 0$$ 
In this example, to serve as a good IV, $selectyrs$ needs to satisfy the assumptions that it is uncorrelated with the residual, and correlated with $choiceyrs$.

(In our model, the variables $black$,$hispanic$, $female$ are exogenous variables, $choiceyrs$ is the endogenous varible and $selectyrs$ is IV.)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 4</h3>

In [7]:
res_s2 = smf.ols('mnce ~ black + hispanic + female + fitted', data=df).fit(cov_type = 'HC3')
print(res_s2.summary())

                            OLS Regression Results                            
Dep. Variable:                   mnce   R-squared:                       0.086
Model:                            OLS   Adj. R-squared:                  0.082
Method:                 Least Squares   F-statistic:                     19.81
Date:                Sat, 07 Dec 2019   Prob (F-statistic):           1.07e-15
Time:                        12:39:15   Log-Likelihood:                -4367.8
No. Observations:                 990   AIC:                             8746.
Df Residuals:                     985   BIC:                             8770.
Df Model:                           4                                         
Covariance Type:                  HC3                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     57.0680      1.890     30.190      0.0

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
"Second stage" equation: 
$ mnce = 57.0680 - 0.2413 \widehat{choiceyrs} - 16.3169 black - 13.7754 hispanic + 1.3197 female$

Compared to the results we've got in part 1, the p-value on $\widehat{choiceyrs}$ does not decrease, in fact, its larger than before, which indicates that the $selectyrs$ might not be a good IV in this case.

The reason for the unsatisfied results that is $selectyrs$ is not significant in the original model as we did in part(1), from this IV model, we are more centain that $selectyrs$ is not a good explanatory variable for $mnce$, thus in further analysis, we need to drop $selectyrs$ in our original model. This is not a good case, usually we are interested in analyzing the endogenous variable which is significant in OLS, but the 2SLS method here is what you need to know.


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 5</h3>

In [8]:
import linearmodels.iv as iv  

In [9]:
res_iv = iv.IV2SLS.from_formula('mnce ~ 1 + black + hispanic + female + [choiceyrs ~ selectyrs]', df).fit()
print(res_iv.summary)

                          IV-2SLS Estimation Summary                          
Dep. Variable:                   mnce   R-squared:                      0.0864
Estimator:                    IV-2SLS   Adj. R-squared:                 0.0827
No. Observations:                 990   F-statistic:                    80.276
Date:                Sat, Dec 07 2019   P-value (F-stat)                0.0000
Time:                        12:39:16   Distribution:                  chi2(4)
Cov. Estimator:                robust                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      57.068     1.8772     30.400     0.0000      53.389      60.747
black         -16.317     1.9476    -8.3779     0.00

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>alternative method</h3>

In [10]:
from statsmodels.api import add_constant
from linearmodels.iv import IV2SLS

In [11]:
df = add_constant(df, has_constant='add')

C:\Users\Owner\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [12]:
dep = 'mnce'
exog = ['const','black','hispanic','female']
endog = ['choiceyrs']
instr = ['selectyrs']

In [13]:
reg4 = IV2SLS(df[dep],df[exog],df[endog],df[instr]).fit(cov_type='unadjusted')
print(reg4)

#reg4 = IV2SLS(df.mnce, df[['const','black','hispanic','female']], df.choiceyrs, df.selectyrs).fit(cov_type='unadjusted')

                          IV-2SLS Estimation Summary                          
Dep. Variable:                   mnce   R-squared:                      0.0864
Estimator:                    IV-2SLS   Adj. R-squared:                 0.0827
No. Observations:                 990   F-statistic:                    93.054
Date:                Sat, Dec 07 2019   P-value (F-stat)                0.0000
Time:                        12:39:19   Distribution:                  chi2(4)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const          57.068     1.6535     34.513     0.0000      53.827      60.309
black         -16.317     1.8102    -9.0137     0.00

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
There are two problems if we manually run 2SLS as we did in part 2 and 4. 1)It's tedious. 2)The standard errors are not correct. Both of these problems are solved by using a two state least squares estimator. 

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1>Question 2</h1>

There is a survey of House of Representatives elections in 1988 and 1990. Only winners from 1988 who are also running in 1990 appear in the sample; these are the incumbents. For the years 1988 and 1990, an unobserved effects model explaining the share of the incumbent’s vote in terms of expenditures by both candidates is
$$vote_{it} = \beta_0 + \delta_0 d90_t + \beta_1 log(inexp_{it}) + \beta_2 log(chexp_{it}) + \beta_3 incshr_{it} + a_i + u_{it}, t = 1,2,$$
where  
$d90_t$ is a dummy variable for 1990;  
$inexp_{it}$ is the incumbent $i$’s spending in year $i$, in the dataset, they are $inexp88$ and $inexp90$ in 1988 and 1990 respectively;  
$chexp_{it}$ is the challenger $i$'s spending in year $i$, in the dataset, they are $chexp88$ and $chexp90$ in 1988 and 1990 respectively;   
$incshr_{it}$ is the incumbent $i$’s share of total campaign spending in year $i$ (in percentage form), eg. 
$incshr90 = 100 * {inexp90 \over {inexp90+chexp90}}$;  
$a_i$ is the unobserved effect and contains characteristics of the incumbent $i$ — such as “quality” — as well as things about the district that are constant. The incumbent $i$’s gender and party are constant over time, so these are subsumed in $a_i$.  

Before we want to study further of the model, let's firstly use the dataset below the construct the model.

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

In [2]:
df3 = pd.read_csv('https://github.com/MariaMeng22/econometric-001/raw/master/vote_new.csv')
df3.head()

state  district   democ  vote90  vote88  inexp90  chexp90  inexp88 chexp88  \
0  "AL"          2      0      51      94   596096   163663   234923       .   
1  "AL"          3      1      74      65   176550    22989   679297  443927   
2  "AL"          7      1      71      68   238446    58952   328296    8737   
3  "AK"          1      0      52      62   564759   164732   626377  402477   
4  "AZ"          2      1      66      73   112373     1445    99607    3065   

  tenure lawyer  win88  win90  rptchall  
0     26      1      1      1         0  
1      1      0      1      1         0  
2      4      1      1      1         0  
3     17      0      1      1         0  
4     29      1      1      1         1

(1) Generate $log(inexp_{it}$) and $log(chexp_{it})$ from $inexp_{it}$ and $chexp_{it}$. For example, in the dataset, get $linexp90$ by taking the logarithm of $inexp90$.

(2) Generate $incshr_{it}$ accoding to its definition, name it to $incshr88$ and $incshr90$ for 1988 and 1990 respectively.  

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 1</h3>

In [3]:
#Before dropping null, we need to drop those variables we won't use which also contain null
df3 = df3.drop(['tenure', 'lawyer'], axis = 1)
#df3 = df3[['state', 'district','democ', 'vote90', 'vote88', 'inexp90', 'chexp90','inexp88', 'chexp88', 'win88', 'win90', 'rptchall']]

In [4]:
#drop the null for our variables in use
df3 = df3.replace('\.+', np.nan, regex=True).dropna()
#check the sample size after dropping
df3.shape

(157, 12)

In [5]:
#Before doing computations among columns, we first need to convert column type into float
df3[['inexp88', 'inexp90','chexp88','chexp90']] = \
df3[['inexp88', 'inexp90','chexp88','chexp90']].astype(float)

In [6]:
#Add computed columns
df3['linexp90']=np.log(df3['inexp90'].astype('float64'))
df3['linexp88']=np.log(df3['inexp88'].astype('float64'))
df3['lchexp88']=np.log(df3['chexp88'].astype('float64'))
df3['lchexp90']=np.log(df3['chexp90'].astype('float64'))
df3.head()

state  district   democ  vote90  vote88   inexp90   chexp90   inexp88  \
1  "AL"          3      1      74      65  176550.0   22989.0  679297.0   
2  "AL"          7      1      71      68  238446.0   58952.0  328296.0   
3  "AK"          1      0      52      62  564759.0  164732.0  626377.0   
4  "AZ"          2      1      66      73  112373.0    1445.0   99607.0   
5  "AZ"          3      0      57      69  225149.0    9353.0  319690.0   

    chexp88  win88  win90  rptchall   linexp90   linexp88   lchexp88  \
1  443927.0      1      1         0  12.081359  13.428814  13.003415   
2    8737.0      1      1         0  12.381898  12.701671   9.075322   
3  402477.0      1      1         0  13.244154  13.347708  12.905393   
4    3065.0      1      1         1  11.629579  11.508988   8.027803   
5   26281.0      1      1         0  12.324518  12.675107  10.176602   

    lchexp90  
1  10.042771  
2  10.984479  
3  12.012075  
4   7.275865  
5   9.143452

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 2</h3>

In [7]:
df3['incshr90'] = 100 * df3['inexp90'] / (df3['inexp90'] + df3['chexp90']).astype('float64')
df3['incshr88'] = 100 * df3['inexp88'] / (df3['inexp88'] + df3['chexp88']).astype('float64')

In [8]:
df3.head()

state  district   democ  vote90  vote88   inexp90   chexp90   inexp88  \
1  "AL"          3      1      74      65  176550.0   22989.0  679297.0   
2  "AL"          7      1      71      68  238446.0   58952.0  328296.0   
3  "AK"          1      0      52      62  564759.0  164732.0  626377.0   
4  "AZ"          2      1      66      73  112373.0    1445.0   99607.0   
5  "AZ"          3      0      57      69  225149.0    9353.0  319690.0   

    chexp88  win88  win90  rptchall   linexp90   linexp88   lchexp88  \
1  443927.0      1      1         0  12.081359  13.428814  13.003415   
2    8737.0      1      1         0  12.381898  12.701671   9.075322   
3  402477.0      1      1         0  13.244154  13.347708  12.905393   
4    3065.0      1      1         1  11.629579  11.508988   8.027803   
5   26281.0      1      1         0  12.324518  12.675107  10.176602   

    lchexp90   incshr90   incshr88  
1  10.042771  88.478944  60.477429  
2  10.984479  80.177405  97.407672  
3  12.012075  77.418227  60.881039  
4   7.275865  98.730429  97.014765  
5   9.143452  96.011548  92.403699

According to the model we constructed, we are interested in the effect of campaign expenditures on election outcomes.   

(3) Difference the given equation across the two years and estimate the differenced equation by OLS. Which variables are individually significant at the 5% level (you can name the differenced variable as you want)?  

(4) In the equation from part (3), test for joint significance of the explanatory variables. Report the p-value.  <span style="color:red">(That is, the overal significance of a regression, which is a test of the joint significance of all explanatory variables appearing in a multiple regression equation, test for overall significance is a special case of test for joint significance, see more in the solution. )

(5) Reestimate the equation from part (3) using $\Delta incshr$ as the only independent variable. Interpret the coefficient on $\Delta incshr$. 

(6) Redo part (3), but now use only the pairs that have repeat challengers. (This allows us to control for characteristics of the challengers as well, which would be in $a_i$)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 3</h3>

In [22]:
df3['cvote'] = df3['vote90'] - df3['vote88'] 
df3['clinexp'] = df3['linexp90'] - df3['linexp88'] 
df3['clchexp'] = df3['lchexp90'] - df3['lchexp88'] 
df3['cincshr'] = df3['incshr90'] - df3['incshr88'] 
df3['cvote'] = df3['vote90'] - df3['vote88']

In [23]:
reg = smf.ols('cvote ~ clinexp + clchexp + cincshr',data=df3).fit()
print(reg.summary())

                            OLS Regression Results                            
Dep. Variable:                  cvote   R-squared:                       0.244
Model:                            OLS   Adj. R-squared:                  0.229
Method:                 Least Squares   F-statistic:                     16.43
Date:                Sat, 07 Dec 2019   Prob (F-statistic):           2.63e-09
Time:                        12:39:22   Log-Likelihood:                -541.49
No. Observations:                 157   AIC:                             1091.
Df Residuals:                     153   BIC:                             1103.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -2.5559      0.631     -4.051      0.0

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
For the year 1988, the unobserved effects model is
$vote_{i,1988} = \beta_0 + \beta_1 log(inexp_{i,1988}) + \beta_2  log(chexp_{i,1988}) + \beta_3 incshr_{i,1988} + \alpha_i + u_{i,1988} $
For the year 1990, the unobserved effects model is
$vote_{i,1990} = \beta_0 + \delta_0 + \beta_1 log(inexp_{i,1990}) + \beta_2  log(chexp_{i,1990}) + \beta_3 incshr_{i,1990} + \alpha_i + u_{i,1990} $
On differencing the given equation across the two years, the model is
    $\Delta vote_i = \delta_0 + \beta_1 \Delta log(inexp_i) + \beta_2 \Delta log(chexp_i) + \beta_3 \Delta incshr_i + \Delta u_i $

The coefficient of the explanatory variable $\Delta incshr$ has the p-value .0155 which is less than the critical p-value of 0.05 at 5% level of significance, thereby, indicating that this variable is individually significant at the 5% level of significance. 

Rest of the explanatory variables is not statistically significant astheir respective p-value is greater than the critical p-value of 0.05 at 5% level of significance.

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 4</h3>

F-statistics in the results output is 16.43, p-value is approximately to 0, that is, the model passes the test for overall significance at 0.05 level of significance.

<span style="color:red">If you want to test the overall significance, the formula is
    
<span style="color:red">$F_{k,N-k-1}$ = $R^2/k \over (1 - R^2)/(N-k-1)$

<span style="color:red">If you want to test the joint significance of some of explanatory varaibles, not all, for exmaple, as in my original question, to test for joint significance of $clinexp$ and $clchexp$, the formula is

<span style="color:red">$F_{q,N-k-1}$ = $(R^2_R - R^2_{UR})/q \over R^2_{UR}/(N-k-1)$

<span style="color:red">See an example https://tien89.wordpress.com/tag/jointly-significant/ for reference.

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 5</h3>

In [24]:
reg2 = smf.ols('cvote ~ cincshr',data=df3).fit()
print(reg2.summary())

                            OLS Regression Results                            
Dep. Variable:                  cvote   R-squared:                       0.229
Model:                            OLS   Adj. R-squared:                  0.224
Method:                 Least Squares   F-statistic:                     45.97
Date:                Sat, 07 Dec 2019   Prob (F-statistic):           2.38e-10
Time:                        12:39:22   Log-Likelihood:                -543.02
No. Observations:                 157   AIC:                             1090.
Df Residuals:                     155   BIC:                             1096.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -2.6811      0.625     -4.288      0.0

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
The coefficient of $\Delta incshr$ is 0.2176, indicating that a 10% increase in the incumbent's share of total spending $incshr$ leads to 2.18% increase in incumbent's vote.

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h3>part 6</h3>

In [25]:
#c 
df3_rptchall = df3[(df3['rptchall'] == 1)]
df3_rptchall.head()

state  district   democ  vote90  vote88   inexp90   chexp90   inexp88  \
4   "AZ"          2      1      66      73  112373.0    1445.0   99607.0   
10  "CA"          3      1      60      71  734005.0    4628.0  638688.0   
16  "CA"         11      1      66      71  620782.0   97030.0  269510.0   
21  "CA"         23      1      62      64  201404.0  360389.0  140486.0   
22  "CA"         24      1      69      72  287505.0    1830.0  191334.0   

     chexp88  win88  ...   linexp90   linexp88   lchexp88   lchexp90  \
4     3065.0      1  ...  11.629579  11.508988   8.027803   7.275865   
10    7695.0      1  ...  13.506271  13.367171   8.948326   8.439880   
16   95575.0      1  ...  13.338735  12.504361  11.467667  11.482775   
21  100956.0      1  ...  12.213068  11.852863  11.522440  12.794939   
22   15449.0      1  ...  12.568996  12.161776   9.645300   7.512071   

     incshr90   incshr88  cvote   clinexp   clchexp    cincshr  
4   98.730429  97.014765     -7  0.120591 -0.751938   1.715664  
10  99.373437  98.809529    -11  0.139100 -0.508446   0.563908  
16  86.482533  73.821165     -5  0.834374  0.015109  12.661368  
21  35.850215  58.186231     -2  0.360205  1.272499 -22.336016  
22  99.367515  92.528883     -3  0.407220 -2.133228   6.838632  

[5 rows x 22 columns]

In [26]:
reg3 = smf.ols('cvote ~ cincshr',data=df3_rptchall).fit()
print(reg3.summary())

                            OLS Regression Results                            
Dep. Variable:                  cvote   R-squared:                       0.037
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     1.189
Date:                Sat, 07 Dec 2019   Prob (F-statistic):              0.284
Time:                        12:39:22   Log-Likelihood:                -103.08
No. Observations:                  33   AIC:                             210.2
Df Residuals:                      31   BIC:                             213.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -2.2498      0.999     -2.253      0.0

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
The coefficient of $\Delta incshr$ is 0.0924, indicating that a 10% increase in the incubent's share of total spending leads to 0.924% increase in incubent's vote.